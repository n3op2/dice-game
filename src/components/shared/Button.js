import React from 'react';
import AppContext from './AppContext';

const Button = ({ onClick, name = 'Unknown' }) => {
    if (!onClick) throw new Error('missing onClick prop');

    const { playing } = React.useContext(AppContext);
    const isDisabled = typeof playing === undefined ? false : playing;

    return (
        <button className="button-1" disabled={isDisabled} onClick={onClick}>
            {name}
        </button>
    );
};

export default Button;
