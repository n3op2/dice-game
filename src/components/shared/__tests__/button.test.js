import React from 'react';
import { shallow } from 'enzyme';
import Button from '../Button';

const Component = <Button onClick={console.log} name="test" />;

describe('Text Component', () => {
    let wrapper;

    beforeEach(() => {
        React.useContext = jest.fn().mockReturnValue({ rolling: true });
        wrapper = shallow(Component);
    });

    afterEach(jest.clearAllMocks);

    describe('invalid or missing props', () => {
        beforeEach(() => {
            React.useContext = jest.fn().mockReturnValue('none');
        });

        test('renders even if no props loaded from context', () => {
            expect(shallow(Component)).toMatchSnapshot();
        });

        test('throws if onClick function is not provided', () => {
            expect(() => shallow(<Button />)).toThrowError(
                'missing onClick prop'
            );
        });

        test('renders with "Unknown" if name props is not provided', () => {
            expect(() =>
                shallow(<Button onClick={console.log} />)
            ).toMatchSnapshot();
        });
    });

    describe('if state is rolling set to true', () => {
        test('renders a disabled component', () => {
            React.useContext = jest.fn().mockReturnValue({ rolling: false });
            expect(shallow(Component)).toMatchSnapshot();
        });
    });

    test('gets context (state)', () => {
        expect(React.useContext).toHaveBeenCalledWith(expect.toMatchSnapshot());
    });

    test('renders a component', () => {
        expect(wrapper).toMatchSnapshot();
    });
});
