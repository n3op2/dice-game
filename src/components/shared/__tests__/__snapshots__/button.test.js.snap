// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component gets context (state) 1`] = `
Object {
  "$$typeof": Symbol(react.context),
  "Consumer": Object {
    "$$typeof": Symbol(react.context),
    "_calculateChangedBits": null,
    "_context": [Circular],
  },
  "Provider": Object {
    "$$typeof": Symbol(react.provider),
    "_context": [Circular],
  },
  "_calculateChangedBits": null,
  "_currentRenderer": null,
  "_currentRenderer2": null,
  "_currentValue": undefined,
  "_currentValue2": undefined,
  "_threadCount": 0,
}
`;

exports[`Text Component if state is rolling set to true renders a disabled component 1`] = `
<button
  className="button-1"
  onClick={[Function]}
>
  test
</button>
`;

exports[`Text Component invalid or missing props renders even if no props loaded from context 1`] = `
<button
  className="button-1"
  onClick={[Function]}
>
  test
</button>
`;

exports[`Text Component invalid or missing props renders with "Unknown" if name props is not provided 1`] = `[Function]`;

exports[`Text Component renders a component 1`] = `
<button
  className="button-1"
  onClick={[Function]}
>
  test
</button>
`;
